formats/json-okio/commonMain/src/kotlinx/serialization/json/okio/internal/OkioJsonStreams.kt

```diff
@@ -42,7 +42,7 @@ internal class JsonToOkioStreamWriter(private val target: BufferedSink) : JsonWr
     }
 
     override fun release() {
-        target.flush()
+        target.emitCompleteSegments()
```

Comment:
Since you are not wrapping the sink and instead are accepting an already-buffered copy, there is no need to call any flush/emit/emitCompleteSegment function at the end. Bytes will always be fully available in the caller's control.

So this call (or the whole release concept) can be removed.

Reply:
A `BufferedSink` will automatically call `emitCompleteSegments` after each granular call to a `write*` function. So this call will always be a no-op as the previous `write*` will have already invoked it and sent any complete segments to the underlying `Sink`.

The only time you would get runaway memory usage is if you wrote directly to the `BufferedSink`'s `Buffer` rather than to the `BufferedSink` itself. You would only do this if for some reason you wanted to enqueue more bytes than a single segment before emitting them a single unit.

So a user doing a loop of encodes will see segment-sized writes automatically and then they can choose whether they want to add a call to `emit()` inside their loop or if no new elements show up (or whatever policy they want).

Generally I recommend that the person who creates the `BufferedSink` should be the one responsible for any calls to `flush`/`emit`/`emitCompleteSegments` because they have the application-specific knowledge to the behavior they want (in addition to the automatic internal calls to `emitCompleteSegments`).

Reply:
The serializer uses the functions `writeUtf8` and `writeUtf8CodePoint` - they do not call either `emit` or `emitCompleteSegments` inside

Reply:
The implementation of those functions on `BufferedSink` are as follows.

* `writeUtf8`: https://github.com/square/okio/blob/ad6f80d954bdd727f9757f85d4109d7b42665604/okio/c/commonMain/kotlin/okio/internal/-RealBufferedSink.kt#L52-L56
* `writeUtf8CodePoint`: https://github.com/square/okio/blob/ad6f80d954bdd727f9757f85d4109d7b42665604/okio/src/commonMain/kotlin/okio/internal/-RealBufferedSink.kt#L68-L72

Both write to the backing `Buffer` and then transfer any complete segments from that buffer to the underlying `Sink`.

Reply:
Won't this cause problems with custom implementations of `BufferedSink`? Because we can be sure that `emit...` is called only for `RealBufferedSink`, however, we accept any inheritors of `BufferedSink` in the function.

Reply:
Perhaps it should be indicated in the documentation that the users themselves should monitor the buffer and explicitly emit it to the `Sink` if they use a custom implementation?
Then there really won't be any sense in call `emit...` in the serialization code.

Reply:
Realistically there are no third-party subtypes of the `Buffered*` interfaces. The interfaces should be sealed now that such a thing is possible in Kotlin. `Sink` is the type you create and then you use the library to buffer it for you. 

`BufferedSink` would be a class with a private ctor but `Buffer` needs to be both a `BufferedSink` and `BufferedSource` which necessitated the interface usage.

Leaving this emit call certainly won't harm anything, but in practice it will 100% be a no-op. Sticking with the comparison to Moshi and Wire that got us to the point about accepting a `BufferedSink`, you'll notice that they do not contain such calls either.